Feature/mapo running script

thiagopbueno opened the pull request:
Scaffold script for calling mapo trainers with configuration given by command-line arguments. Needs some work to make experiments more practical by making better use of [Tune](https://ray.readthedocs.io/en/latest/tune-usage.html#training-features), but it's a start.

Overall, the main goal here is to make sure all components (custom envs, trainers, policies) are  smoothly integrated (and also for me to gain experience with rllib training outputs, metrics reported, ...)

0xangelo committed:
feat: add full bootstrap on-policy actor loss

Signed-off-by: Ângelo Lovatto <angelolovatto@gmail.com>

0xangelo committed:
refactor: make original function arguments explicit

Signed-off-by: Ângelo Lovatto <angelolovatto@gmail.com>

0xangelo committed:
refactor: make loss computation more functional

Signed-off-by: Ângelo Lovatto <angelolovatto@gmail.com>

thiagopbueno committed:
feat(envs): register custom gym envs in Tune

closes #49

Signed-off-by: Thiago P. Bueno <thiago.pbueno@gmail.com>

thiagopbueno committed:
feat(envs): register custom gym envs in Tune

closes #49

Signed-off-by: Thiago P. Bueno <thiago.pbueno@gmail.com>

thiagopbueno committed:
feat(script): add MAPO running script with optional arguments

closes #37

Signed-off-by: Thiago P. Bueno <thiago.pbueno@gmail.com>

thiagopbueno committed:
feat(script): add MAPO running script with optional arguments

closes #37

Signed-off-by: Thiago P. Bueno <thiago.pbueno@gmail.com>

thiagopbueno committed:
fix(test): import NavigationEnv from tune.registry in test_navigation.py

Signed-off-by: Thiago P. Bueno <thiago.pbueno@gmail.com>

thiagopbueno committed:
fix(test): import NavigationEnv from tune.registry in test_navigation.py

Signed-off-by: Thiago P. Bueno <thiago.pbueno@gmail.com>

0xangelo committed:
refactor: standardize model inputs

Also remove tests that checked if policy and Q models were serializable.
These were freezing and didn't test for a necessary functionality.

Signed-off-by: Ângelo Lovatto <angelolovatto@gmail.com>

0xangelo committed:
fix(mapo): initialize dynamics variables on model init

Signed-off-by: Ângelo Lovatto <angelolovatto@gmail.com>

0xangelo committed:
test: add tests for MAPOModel and GaussianDynamicsModel

Signed-off-by: Ângelo Lovatto <angelolovatto@gmail.com>

0xangelo commented on scripts/mapo:
```diff
@@ -0,0 +1,231 @@
+        # resources_per_trial={
+        #     "cpu": args.num_cpus,
+        #     "gpu": args.num_gpus,
+        # },
```
Should we keep this comment here? Is it necessary?

thiagopbueno replied:
There's an issue here actually. If I uncomment those I get the following error:

```bash
Traceback (most recent call last):
  File "/home/tbueno/Projects/model-based-reinforcement-learning/bin/mapo", line 7, in <module>
    exec(compile(f.read(), __file__, 'exec'))
  File "/home/tbueno/Projects/model-based-reinforcement-learning/model-aware-policy-optimization/scripts/mapo", line 230, in <module>
    results = run(arguments)
  File "/home/tbueno/Projects/model-based-reinforcement-learning/model-aware-policy-optimization/scripts/mapo", line 200, in run
    "dynamics": args.config_dynamics_net,
  File "/home/tbueno/Projects/model-based-reinforcement-learning/lib/python3.7/site-packages/ray/tune/tune.py", line 244, in run
    runner.step()
  File "/home/tbueno/Projects/model-based-reinforcement-learning/lib/python3.7/site-packages/ray/tune/trial_runner.py", line 323, in step
    next_trial = self._get_next_trial()  # blocking
  File "/home/tbueno/Projects/model-based-reinforcement-learning/lib/python3.7/site-packages/ray/tune/trial_runner.py", line 486, in _get_next_trial
    self._update_trial_queue(blocking=wait_for_trial)
  File "/home/tbueno/Projects/model-based-reinforcement-learning/lib/python3.7/site-packages/ray/tune/trial_runner.py", line 632, in _update_trial_queue
    trials = self._search_alg.next_trials()
  File "/home/tbueno/Projects/model-based-reinforcement-learning/lib/python3.7/site-packages/ray/tune/suggest/basic_variant.py", line 57, in next_trials
    trials = list(self._trial_generator)
  File "/home/tbueno/Projects/model-based-reinforcement-learning/lib/python3.7/site-packages/ray/tune/suggest/basic_variant.py", line 87, in _generate_trials
    experiment_tag=experiment_tag)
  File "/home/tbueno/Projects/model-based-reinforcement-learning/lib/python3.7/site-packages/ray/tune/config_parser.py", line 189, in create_trial_from_spec
    **trial_kwargs)
  File "/home/tbueno/Projects/model-based-reinforcement-learning/lib/python3.7/site-packages/ray/tune/trial.py", line 306, in __init__
    trainable_cls, default_resources))
ValueError: Resources for <class 'ray.rllib.agents.trainer_template.MAPO'> have been automatically set to Resources(cpu=1, gpu=0, extra_cpu=0, extra_gpu=0, custom_resources={}, extra_custom_resources={}) by its `default_resource_request()` method. Please clear the `resources_per_trial` option.
```

Do you have any idea on how to fix it?

0xangelo replied:
It seems that the `resources_per_trial` option is overridden by the [trainer config](https://github.com/ray-project/ray/blob/e4854b11f9a032261a3a07029b8492e59002da61/python/ray/rllib/agents/trainer.py#L132):
```python
    # === Resources ===
    # Number of actors used for parallelism
    "num_workers": 2,
    # Number of GPUs to allocate to the driver. Note that not all algorithms
    # can take advantage of driver GPUs. This can be fraction (e.g., 0.3 GPUs).
    "num_gpus": 0,
    # Number of CPUs to allocate per worker.
    "num_cpus_per_worker": 1,
    # Number of GPUs to allocate per worker. This can be fractional.
    "num_gpus_per_worker": 0,
    # Any custom resources to allocate per worker.
    "custom_resources_per_worker": {},
    # Number of CPUs to allocate for the driver. Note: this only takes effect
    # when running in Tune.
    "num_cpus_for_driver": 1,
```
So we should pass the command line options as the `num_cpus_for_driver` and `num_gpus`  configurations.

thiagopbueno replied:
I've just tried that and it didn't seem to solve the issue...

0xangelo replied:
I tried it with the following command and it worked at the time:
```shell
mapo --run MAPO --env Navigation-v0 --config-actor-net actor.json --config-critic-net critic.json --config-dynamics-net dynamics.json
```

thiagopbueno replied:
Indeed, just needed to merge your feature/on-policy-madpg branch onto my local feature/mapo-running-script.. and issue solved!

Thanks!

thiagopbueno replied:
Sorry, I made a confusion here... actually the problem of the num_gpus and num_cpus persists...

I managed to solve the other issue on the `critic_delay` argument (solved by your commit 39d4d13)... but nevermind we can check this together this week.

0xangelo commented on scripts/mapo:
```diff
@@ -0,0 +1,231 @@
+                    "dynamics": args.config_dynamics_net,
+                },
+            },
+        },
```
Just to be clear: are we purposefully ignoring some command line arguments such as delays and sample batch sizes?

thiagopbueno replied:
I forgot to add those. Just did it in this follow-up commit b377b99

thiagopbueno replied:
But again, there's still an awful lot of arguments to add... this is just a scaffold... also not entirely confident there isn't a better way of passing args from CLI to the trainer's config dict... but let's no dwell too much on details here..

thiagopbueno replied:
BTW, I've just noticed that for ``trainer=MAPO`` passing the argument `critic_delay` in config causes an error:

```bash
$ mapo --run=MAPO --env=Navigation-v0 --config-actor scripts/actor.json --config-critic scripts/critic.json --config-dynamics scripts/dynamics.json --timesteps-total 1000
```

```bash
Traceback (most recent call last):
  File "/home/tbueno/Projects/model-based-reinforcement-learning/lib/python3.7/site-packages/ray/tune/trial_runner.py", line 498, in _process_trial
    result = self.trial_executor.fetch_result(trial)
  File "/home/tbueno/Projects/model-based-reinforcement-learning/lib/python3.7/site-packages/ray/tune/ray_trial_executor.py", line 342, in fetch_result
    result = ray.get(trial_future[0])
  File "/home/tbueno/Projects/model-based-reinforcement-learning/lib/python3.7/site-packages/ray/worker.py", line 2247, in get
    raise value
ray.exceptions.RayTaskError: ray_MAPO:train() (pid=24875, host=system76-ubuntu)
  File "/home/tbueno/Projects/model-based-reinforcement-learning/lib/python3.7/site-packages/ray/rllib/agents/trainer_template.py", line 87, in __init__
    Trainer.__init__(self, config, env, logger_creator)
  File "/home/tbueno/Projects/model-based-reinforcement-learning/lib/python3.7/site-packages/ray/rllib/agents/trainer.py", line 328, in __init__
    Trainable.__init__(self, config, logger_creator)
  File "/home/tbueno/Projects/model-based-reinforcement-learning/lib/python3.7/site-packages/ray/tune/trainable.py", line 99, in __init__
    self._setup(copy.deepcopy(self.config))
  File "/home/tbueno/Projects/model-based-reinforcement-learning/lib/python3.7/site-packages/ray/rllib/agents/trainer.py", line 429, in _setup
    self._allow_unknown_subkeys)
  File "/home/tbueno/Projects/model-based-reinforcement-learning/lib/python3.7/site-packages/ray/tune/util.py", line 172, in deep_update
    raise Exception("Unknown config parameter `{}` ".format(k))
Exception: Unknown config parameter `critic_delay` 
```
Is this related to the fact in `mapo.py` trainer module there is no `critic_delay` in the `DEFAULT_CONFIG` ?

0xangelo replied:
Probably the case, I think I fixed this in my other PR (#50)

thiagopbueno replied:
In which commit / file?

0xangelo replied:
This one 
39d4d13

thiagopbueno replied:
Thanks!

0xangelo commented on scripts/mapo:
```diff
@@ -0,0 +1,231 @@
+
+    print("tensorboard --logdir={}".format(best_logdir))
+    print()
+
```
Looks pretty neat! I like the idea of automatically reporting the best results.

thiagopbueno replied:
Still need to learn how to deal with the [Analysis](https://ray.readthedocs.io/en/latest/tune-package-ref.html?highlight=ExperimentAnalysis#ray.tune.Analysis) object returned by `ray.tune.run()` ...

0xangelo reviewed (commented):
Everything looks fine as it is, just waiting to see if we didn't forget anything in `scripts/mapo` (see my comment.

thiagopbueno committed:
feat(script): pass optimization and exec args to trainer in script/mapo

add train and sample batch sizes, policy and critic delay, and
learning rates for actor, critic and dynamics models' optimizers.

Signed-off-by: Thiago P. Bueno <thiago.pbueno@gmail.com>

thiagopbueno committed:
feat(script): pass optimization and exec args to trainer in script/mapo

add train and sample batch sizes, policy and critic delay, and
learning rates for actor, critic and dynamics models' optimizers.

Signed-off-by: Thiago P. Bueno <thiago.pbueno@gmail.com>

0xangelo committed:
test: increase pure exploration sample size

Signed-off-by: Ângelo Lovatto <angelolovatto@gmail.com>

thiagopbueno committed:
refactor(env): add _transition_fn and reward_fn to NavigationEnv

compute the transition and reward function in TensorFlow

Signed-off-by: Thiago P. Bueno <thiago.pbueno@gmail.com>

thiagopbueno committed:
refactor(env): add _transition_fn and reward_fn to NavigationEnv

compute the transition and reward function in TensorFlow

Signed-off-by: Thiago P. Bueno <thiago.pbueno@gmail.com>

thiagopbueno committed:
Merge branch 'feature/gym-env-navigation' into develop

thiagopbueno committed:
fix(script): change --num-cpus to --num-cpus-for-driver in scripts/mapo

Signed-off-by: Thiago P. Bueno <thiago.pbueno@gmail.com>

thiagopbueno committed:
fix(script): change --num-cpus to --num-cpus-for-driver in scripts/mapo

Signed-off-by: Thiago P. Bueno <thiago.pbueno@gmail.com>

thiagopbueno committed:
Merge branch 'feature/mapo-running-script' of https://github.com/thiagopbueno/model-aware-policy-optimization into feature/mapo-running-script

thiagopbueno committed:
Merge branch 'feature/on-policy-madpg' of https://github.com/thiagopbueno/model-aware-policy-optimization into feature/on-policy-madpg

thiagopbueno committed:
Merge branch 'feature/on-policy-madpg' into feature/mapo-running-script
